sup3r/pipeline/forward_pass.py

```diff
@@ -746,65 +812,77 @@ def run(self):
-                                                      self.lr_pad_slices,
-                                                      self.hr_crop_slices):
-
-                data_chunk = self.data_handler.data[s_low_pad]
-                data[s_high] = ForwardPass.forward_pass_chunk(
-                    data_chunk, crop_slices=s_high_crop,
-                    model_path=self.model_path)
+        if self.cache_file is not None and os.path.exists(self.cache_file):
+            logger.info('Loading forward pass output from cache file: '
+                        f'{self.cache_file}')
+            with open(self.cache_file, 'rb') as f:
+                data = pickle.load(f)
         else:
-            futures = {}
-            now = dt.now()
-            with SpawnProcessPool(
-                    max_workers=self.strategy.pass_workers) as exe:
-                for s_high, s_low_pad, s_high_crop in zip(self.hr_slices,
-                                                          self.lr_pad_slices,
-                                                          self.hr_crop_slices):
-
-                    data_chunk = self.data_handler.data[s_low_pad]
-                    future = exe.submit(ForwardPass.forward_pass_chunk,
-                                        data_chunk=data_chunk,
-                                        crop_slices=s_high_crop,
-                                        model_path=self.model_path)
-                    meta = {'s_high': s_high}
-                    futures[future] = meta
-
-                logger.info(f'Started forward pass for {len(self.hr_slices)} '
-                            f'chunks in {dt.now() - now}.')
-
-                for i, future in enumerate(as_completed(futures)):
-                    slices = futures[future]
-                    data[slices['s_high']] = future.result()
-                    if i % (len(futures) // 10 + 1) == 0:
-                        logger.debug(f'{i+1} out of {len(futures)} forward '
-                                     'passes completed.')
-
-        logger.info('All forward passes are complete.')
-        data = data[:, :, self.ti_hr_crop_slice, :]
+            msg = (f'Starting forward passes on data shape {self.data_shape}. '
+                   f'Using {len(self.lr_slices)} chunks each with shape of '
+                   f'{self.chunk_shape}, spatial_overlap of '
+                   f'{self.strategy.spatial_overlap} and temporal_overlap of '
+                   f'{self.strategy.temporal_overlap}.')
+            logger.info(msg)
+            max_workers = self.max_workers
+            data = np.zeros((self.strategy.s_enhance * self.data_shape[0],
+                             self.strategy.s_enhance * self.data_shape[1],
+                             self.strategy.t_enhance * self.data_shape[2],
+                             len(self.output_features)),
+                            dtype=np.float32)
+            if max_workers == 1:
```

Comment:
This method is wayyy too complex. Consider breaking up into _run_serial() and _run_parallel()